# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/nsHttp.h
# Commit: 373bc48d5325
# Full Hash: 373bc48d5325e17cb67a11035d4542d221023594
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2023-09-19 09:37:28
# Regressor Bug: 1784261
# File Overlap Count: 1
# Description:
#   Bug 1784261 - Add telemetry to understand the reason of connection closure, r=necko-reviewers,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D188132
# ==============================================================================

diff -r f2fee3884997 -r 373bc48d5325 netwerk/protocol/http/nsHttp.h
--- a/netwerk/protocol/http/nsHttp.h	Mon Sep 18 14:28:30 2023 +0000
+++ b/netwerk/protocol/http/nsHttp.h	Mon Sep 18 14:44:18 2023 +0000
@@ -50,6 +50,35 @@
   HTTP_3_VER_1 = 7,
 };
 
+// IMPORTANT: when adding new values, always add them to the end, otherwise
+// it will mess up telemetry.
+enum class ConnectionCloseReason : uint32_t {
+  UNSET = 0,
+  OK,
+  IDLE_TIMEOUT,
+  TLS_TIMEOUT,
+  GO_AWAY,
+  DNS_ERROR,
+  NET_RESET,
+  NET_TIMEOUT,
+  NET_REFUSED,
+  NET_INTERRUPT,
+  NET_INADEQ_SEQURITY,
+  SOCKET_ADDRESS_NOT_SUPPORTED,
+  OUT_OF_MEMORY,
+  SOCKET_ADDRESS_IN_USE,
+  BINDING_ABORTED,
+  BINDING_REDIRECTED,
+  ERROR_ABORT,
+  CLOSE_EXISTING_CONN_FOR_COALESCING,
+  CLOSE_NEW_CONN_FOR_COALESCING,
+  CANT_REUSED,
+  OTHER_NET_ERROR,
+  SECURITY_ERROR,
+};
+
+ConnectionCloseReason ToCloseReason(nsresult aErrorCode);
+
 inline bool IsHttp3(SupportedAlpnRank aRank) {
   return aRank >= SupportedAlpnRank::HTTP_3_DRAFT_29;
 }